File: TreeAppCoreC++/TreeReaders.cpp
#include "pch.h"
#include "TreeReaders.h"

unique_ptr<TreeNode> StaticNamesTreeReader::ConstructTree()
{

	//find root node
	auto rootNodeIterator = find_if(Nodes.begin(), Nodes.end(), [](const StaticNamesNode& s) { return s.ParentId == TreeNode::ROOT_NODE_ID; });

	//construct tree starting from root
	auto rootNode = ConstructTreeNode(rootNodeIterator->Id);

	return rootNode;
}

	unique_ptr<TreeNode> StaticNamesTreeReader::ConstructTreeNode(int nodeId)
	{
		auto node = FindStaticNode(nodeId);

		auto treeNode = make_unique<TreeNode>(node.Id, node.Description);

		for (auto staticNode : Nodes)
		{
			if (staticNode.ParentId == treeNode->Id)
			{
				treeNode->Children.push_back(ConstructTreeNode(staticNode.Id));
			}
		}
		return treeNode;
	}



	StaticNamesNode StaticNamesTreeReader::FindStaticNode(int nodeId)
	{
		//fint the node with Id == nodeId
		auto staticNodeIt = find_if(Nodes.begin(), Nodes.end(),
			[&nodeId](const StaticNamesNode& s)
		{
			return s.Id == nodeId;
		});

		return *staticNodeIt;
	}

	//ex b) ->

	StaticNamesNode StaticNamesTreeReader::FindStaticNodeByName(string nodeName)
	{
		//fint the node with Id == nodeName
		auto staticNodeIt = find_if(Nodes.begin(), Nodes.end(),
			[&nodeName](const StaticNamesNode& s)
			{
				return s.Description == nodeName;
			});

		return *staticNodeIt;
	}

	list<StaticNamesNode> StaticNamesTreeReader::ListNodeSiblings(string nodeName)
	{
		list<StaticNamesNode> result;
		StaticNamesNode target = FindStaticNodeByName(nodeName);


		for (auto j = Nodes.begin(); j != Nodes.end(); j++) {
			if (target.Id != j->Id && target.ParentId == j->ParentId) {

				StaticNamesNode aux = FindStaticNode(j->Id);
				result.push_back(aux);
			}
		}

		return result;
	}

File: TreeAppCore/TreeReaders.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace TreeAppCore
{
    public interface ITreeReader
    {
        TreeNode ConstructTree();
    }

    public class StaticNamesTreeReader : ITreeReader
    {
        private class StaticNamesNode
        {
            public StaticNamesNode(int id, string description, int? parentId = null)
            {
                Id = id;
                Description = description;
                ParentId = parentId;
            }

            public int Id { get; }
            public string Description { get; }
            public int? ParentId { get; }
        };

        private static readonly List<StaticNamesNode> Nodes = 
            new List<StaticNamesNode>
            {
                new StaticNamesNode(10, "Radacina"),
                new StaticNamesNode(100, "Popescu", 10),
                new StaticNamesNode(111, "Papuc", 100),
                new StaticNamesNode(112, "Piedone", 100),
                new StaticNamesNode(113, "Pavel", 100),
                new StaticNamesNode(1131, "Paleologu", 113),
                new StaticNamesNode(113101, "Paleontologu", 1131),
                new StaticNamesNode(113102, "Panageu", 1131),
                new StaticNamesNode(11310201, "Gata", 113102),
                new StaticNamesNode(1132, "Pavaza", 113),
                new StaticNamesNode(200, "Ionescu", 10),
                new StaticNamesNode(221, "Ivan", 200),
                new StaticNamesNode(22101, "Ivanescu", 221),
                new StaticNamesNode(22102, "Ivanciuc", 221),
                new StaticNamesNode(222, "Ioana", 200),
            };

        public TreeNode ConstructTree() {
            //var rootNodeId = Nodes.Where(n => n.ParentId == null).First().Id;

            var rootNodeId = Nodes.SingleOrDefault(n => n.ParentId == null).Id; // Ar trebui sa fie o singura radacina, asa ca aruncam
            var rootNode = ConstructTreeNode(rootNodeId);                      // exceptie in cazul in care sunt mai multe potentiale radacini
            return rootNode;
        }

        private TreeNode ConstructTreeNode(int nodeId)
        {
            var staticNode = FindStaticNode(nodeId);

            var treeNode = new TreeNode { Id = staticNode.Id, Description = staticNode.Description };

            treeNode.Children = Nodes
                .Where(n => n.ParentId == staticNode.Id)
                .Select(n => ConstructTreeNode(n.Id))
                .ToList();  

            return treeNode;
        }

        private StaticNamesNode FindStaticNode(int nodeId)
        {
            return Nodes.SingleOrDefault(n => n.Id == nodeId);
        }
    }
}
